arangod/Network/Methods.cpp

```diff
@@ -259,7 +259,15 @@ void actuallySendRequest(std::shared_ptr<Pack>&& p, ConnectionPool* pool,
           return;
         }
 
-        auto* sch = SchedulerFeature::SCHEDULER;
+        // We access the global SCHEDULER pointer here via an atomic
+        // reference. This is to silence TSAN, which often detects a data
+        // race on this pointer, since this read access here can occasionally
+        // happen before the write in SchedulerFeature::unprepare, which
+        // invalidates the pointer. But even if the read here would happen
+        // later, we check for nullptr below, so all would be good.
+        std::atomic_ref<SupervisedScheduler*> schedulerRef{
+            SchedulerFeature::SCHEDULER};
+        auto* sch = schedulerRef.load(std::memory_order_relaxed);
```

Comment:
LGTM for code, but comments look strange for me.
If null possible when some synchronization needed, it can be explicit like here or via HB relation.
So this "silence TSAN" sounds wrong.
If exist some HB relation that TSAN ignore, feel free to point that.

Reply:
I try to improve my comment.

Reply:
What hb relation exists what tsan doesn't see?

Reply:
I am not claiming that there is an official HB relation which TSAN does not see. I am only saying that it sees none. It might very well be that there exists none. This place in the code is the place which handles an incoming network response. It is entirely possible that a network response comes in when the scheduler is already shut down. Therefore, we check here if the pointer is already nullptr and act accordingly. Usually, this does not happen, not even in tests. But TSAN sees a load **before** the nullptr store and reckons that there is no HB except for some sleep and notices this. This message by TSAN is irrelevant (although technically correct). Therefore we want to silence TSAN here.

Reply:
The wording is probably not ideal because "happens before" has a very specific meaning, but in "this read access here can occasionally happen before the write" it means something different (if we had such a "happens before" order we would not have a data race). The problem is that this read and the write that sets it to null are _unordered_. However, this does not matter in this case since we have the additional null check. So in this case we do not care about order, but just about atomicity. Nevertheless, @MBkkt is right that maybe we should adjust the wording.